### PR TITLE
feature: add way of testing deps of a module

### PR DESCRIPTION
**Pull request by andrewhughes101**
New feature to test all the dependencies of a module. This will test the versions of the dependencies that get installed at the time of `npm install`. Current work in progress tests and documentation to come. Please let me know if this should be intergarted into citgm or citgm-all - help would be appreciated on how to do this

##### Checklist

- [x] `npm test` passes
- [ ] tests are included
- [ ] documentation is changed or added
- [x] contribution guidelines followed
      [here](https://github.com/nodejs/citgm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by andrewhughes101** [comment]
Refs #631 could be used to automate this

File: bin/citgm-deps.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  sha: app.sha,
+  tmpDir: app.tmpDir,
+  customTest: app.customTest,
+  yarn: app.yarn
+};
+
+const runner = new citgm.Tester(mod, options);
+let pack, dependencies;
+
+//Fisrt npm install in a temp directory and grab all the dependencies and their version in semver format
```
Comment (richardlau) [inline bin/citgm-deps.js]:
```suggestion
// First npm install in a temp directory and grab all the dependencies and their version in semver format
```

File: bin/citgm-deps.js [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  const { npm, yarn } = await packageManager.getPackageManagers();
+  runner.npmPath = npm;
+  runner.yarnPath = yarn;
+  pack = mod.split('@')[0]; //The actual package
+  await doInstall();
+  await findDependencies();
+  await cleanup();
+
+  if (dependencies === undefined) {
+    console.log("Module has no dependencies")
```
Comment (richardlau) [inline bin/citgm-deps.js]:
```suggestion
    console.log('Module has no dependencies')
```

File: bin/citgm-deps.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        launch(mod, options);
+      });
+    } else {
+      launch(mod, options);
+    }
+  }
+};
+
+let doInstall = async () => {
+  //Use a child process to npm install in the tmp directory
```
Comment (richardlau) [inline bin/citgm-deps.js]:
```suggestion
  // Use a child process to npm install in the tmp directory
```

File: bin/citgm-deps.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+  return new Promise((resolve, reject) => {
+    const child = spawn('npm', ['i', '--prefix', runner.path, pack], options);
+    child.on('close', (code) => {
+      console.log('exited with code: ', code);
+      if (code > 0) {
+        console.log("Fail on install")
```
Comment (richardlau) [inline bin/citgm-deps.js]:
```suggestion
        console.log('Fail on install')
```

**Comment by andrewhughes101** [comment]
@richardlau I think an alternative way of producing this functionality would be to add an optional version field to the lookup for each module that citgm-all can handle. Then all that `citgm-deps` would do is the npm install and grab the versions of each dependency.

@nodejs/citgm opinions on adding an optional version field to the lookup?

**Comment by BethGriggs** [comment]
ping @nodejs/citgm on behalf of @andrewhughes101

**Comment by MylesBorins** [comment]
I'm not 100% that this is the best way to approach the problem outlined in #631 

Generally packages don't "just work" with CITGM. Usually we have to manually add them to the lookup table including meta data such as version prefix, package manager default, and what platforms that module is flaky on.

If we automate the process of testing all of the dependencies installed I fear we will end up with a ton of false negatives and in turn have to manually add all of those dependencies into the lookup table before we can get accurate results. At that point a CITGM-ALL run will cover all those dependencies.

We could likely use `tags` to similarly filter subsets of modules if we wanted to only test supported modules. A script that automated updating the `lookup.json` with tags that grouped related modules would likely be very helpful for that.

I'm not going to block this, but would like to hear a bit more about how this will be used and how the data we get from the script will be useful before supporting this feature

**Comment by BridgeAR** [comment]
I agree with @MylesBorins that this might end up with a lot of flaky tests in the test suite. It might be a good idea to try to automate a few more things that we require less meta data (e.g., the version prefix and the default package manager could be detected). What platforms a module is flaky on is something difficult to automate though.

I think this is a good idea on a longer term. I am just not sure if we are already at the point that this works as expected.

**Comment by mhdawson** [comment]
@BridgeAR, @MylesBorins when this work was mentioned to me it was in the context of helping projects run test themselves (for example the Express project) which I think would be useful and a good extension of CITGM.  

I'm thinking we should add the functionality to do that in a way that does not affect our existing CITGM runs (ie we won't need to worry about any flakiness) but does let package maintainers experiment with using CITGM to do additional testing.  What do you think?

**Comment by andrewhughes101** [comment]
The plan was to implement this as a new command and not add it to the CITGM runs. This is so it could be used by module maintainers to test their module dependencies using CITGM without affecting existing nodejs CI test runs. 

If there are features that module maintainers could use in CITGM, then visibility of CITGM could be increased, potentially leading to more contributors?

**Pull request closed without merging**